src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -3359,53 +3363,85 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
+                  typedArg0(rhs) match {
+                    case (rhsTyped, tp, partial) => (treeCopy.AssignOrNamedArg(arg, lhs, rhsTyped), NamedType(name, tp), partial)
+                  }
+                case treeInfo.WildcardStarArg(_) =>
+                  typedArg0(arg) match {
+                    case (argTyped, tp, partial) => (argTyped, RepeatedType(tp), partial)
+                  }
+                case _ =>
+                  typedArg0(arg)
+              }
+            }
 
-                arg match {
-                  // scala/bug#8197/scala/bug#4592 call for checking whether this named argument could be interpreted as an assign
-                  // infer.checkNames must not use UnitType: it may not be a valid assignment, or the setter may return another type from Unit
-                  // TODO: just make it an error to refer to a non-existent named arg, as it's far more likely to be
-                  //       a typo than an assignment passed as an argument
-                  case AssignOrNamedArg(lhs@Ident(name), rhs) =>
-                    // named args: only type the righthand sides ("unknown identifier" errors otherwise)
-                    // the assign is untyped; that's ok because we call doTypedApply
-                    typedArg0(rhs) match {
-                      case (rhsTyped, tp) => (treeCopy.AssignOrNamedArg(arg, lhs, rhsTyped), NamedType(name, tp))
-                    }
-                  case treeInfo.WildcardStarArg(_) =>
-                    typedArg0(arg) match {
-                      case (argTyped, tp) => (argTyped, RepeatedType(tp))
-                    }
-                  case _ =>
-                    typedArg0(arg)
-                }
-              }.unzip
+            val (args1, argTpes, maybeRetypeArgs) = context.savingUndeterminedTypeParams() {
+              val amode = forArgMode(fun, mode)
+              map2(args, altArgPts) { (arg, argPtAlts) => typedSingleArg(arg, argPtAlts, amode, matchAsPartial = true) }.unzip3
             }
             if (context.reporter.hasErrors)
               setError(tree)
             else {
               inferMethodAlternative(fun, undetparams, argTpes, pt)
-              doTypedApply(tree, adaptAfterOverloadResolution(fun, mode.forFunMode, WildcardType), args1, mode, pt)
+              val args2 = if(maybeRetypeArgs contains true) {
+                // There were pattern-matching anonymous functions in the argument list which got typed as PartialFunction
+                // but if the inferred alternative does not actually need a PartialFunction we retype it as Function1.
+                def paramTypes(tpe: Type): List[Type] = tpe match {
+                  case ExistentialType(_, qtpe) => paramTypes(qtpe)
+                  case MethodType(syms, _) => syms.map(_.tpe)
+                  case TypeRef(_, _, tpes) => tpes.init
+                  case _ => Nil
+                }
+                val selectedParams = paramTypes(normalize(fun.tpe))
```

Comment:
`ClassArgsTypeRef` shows up in the last two test cases (passing a case literal to a method with mixed overloads) in `run/InferOverloadedPartialFunction.scala`, both for `fun.tpe` and `fun.symbol.tpe`:

```scala
  assert(ys.collect { case (k, v) => 1 } == 0)
  assert(ys.collect { case (k, v) => (1, 1) } == 2)
```

I wasn't able to reproduce the `ExistentialType` anymore with `partest pos neg run`. Maybe it would show up in a full test run.